plugins/inputs/huebridge/bridgemetadata.go

```diff
@@ -0,0 +1,155 @@
+	}
+	responseData := getResourcesResponse.JSON200.Data
+	if responseData == nil {
+		return make(map[string]string), nil
+	}
+	tree := make(map[string]string, len(*responseData))
+	for _, resource := range *responseData {
+		if resource.Owner != nil {
+			tree[*resource.Id] = *resource.Owner.Rid
+		}
+	}
+	return tree, nil
+}
+
+func fetchDeviceNames(bridgeClient hue.BridgeClient) (map[string]string, error) {
+	getDevicesResponse, err := bridgeClient.GetDevices()
+	if err != nil {
+		return nil, fmt.Errorf("failed to access bridge devices on %q (cause: %w)", bridgeClient.Url().Redacted(), err)
+	}
+	if getDevicesResponse.HTTPResponse.StatusCode != http.StatusOK {
+		return nil, fmt.Errorf("failed to fetch bridge devices from %q (status: %s)", bridgeClient.Url().Redacted(), getDevicesResponse.HTTPResponse.Status)
+	}
+	responseData := getDevicesResponse.JSON200.Data
+	if responseData == nil {
+		return make(map[string]string), nil
+	}
+	names := make(map[string]string, len(*responseData))
+	for _, device := range *responseData {
+		names[*device.Id] = *device.Metadata.Name
+	}
+	return names, nil
+}
+
+func fetchRoomAssignments(bridgeClient hue.BridgeClient) (map[string]string, error) {
+	getRoomsResponse, err := bridgeClient.GetRooms()
+	if err != nil {
+		return nil, fmt.Errorf("failed to access bridge rooms on %q (cause: %w)", bridgeClient.Url().Redacted(), err)
+	}
+	if getRoomsResponse.HTTPResponse.StatusCode != http.StatusOK {
+		return nil, fmt.Errorf("failed to fetch bridge rooms from %q (status: %s)", bridgeClient.Url().Redacted(), getRoomsResponse.HTTPResponse.Status)
+	}
+	responseData := getRoomsResponse.JSON200.Data
+	if responseData == nil {
+		return make(map[string]string), nil
+	}
+	assignments := make(map[string]string, len(*responseData))
+	for _, roomGet := range *responseData {
+		for _, children := range *roomGet.Children {
+			assignments[*children.Rid] = *roomGet.Metadata.Name
+		}
+	}
+	return assignments, nil
+}
+
+func (metadata *bridgeMetadata) resolveResourceRoom(resourceId string, resourceName string) string {
+	roomName := metadata.roomAssignments[resourceName]
+	// If resource does not have a room assigned directly, iterate upwards via
+	// its owners until we find a room or there is no more owner. The latter
+	// may happen (e.g. for Motion Sensors) resulting in room name
+	// "<unassigned>".
+	if roomName == "" {
+		// No room so far, search via the the owner hierarchy...
+		currentResourceId := resourceId
+		for {
+			// Try next owner
+			currentResourceId = metadata.resourceTree[currentResourceId]
+			if currentResourceId == "" {
+				// No owner, no room
+				break
+			}
+			roomName = metadata.roomAssignments[currentResourceId]
+			if roomName != "" {
+				// Room name found, done
+				break
+			}
+		}
+	}
+	if roomName == "" {
+		roomName = "<unassigned>"
+	}
+	return roomName
```

Comment:
How about this shorter version
```suggestion
	if r, found := metadata.roomAssignments[resourceName]; found {
		return r
	}
		
	// If resource does not have a room assigned directly, iterate upwards via
	// its owners until we find a room or there is no more owner. The latter
	// may happen (e.g. for Motion Sensors) resulting in room name
	// "<unassigned>".
	currentResourceId := resourceId
	for {
		// Try next owner
		currentResourceId = metadata.resourceTree[currentResourceId]
		if currentResourceId == "" {
			// No owner left but no room found
			break
		}

		if r := metadata.roomAssignments[currentResourceId]; r != "" {
			// Room name found, done
			return r
		}
	}

	return "<unassigned>"
```

Reply:
Adapted with commit a1e25bfc9f483563b7508b2094f8103d899114e7